drv/lpc55-update-server/src/main.rs

```diff
@@ -226,6 +234,203 @@ impl idl::InOrderUpdateImpl for ServerImpl {
+                // flash layout spreadsheet):
+                //
+                // Page     Addr        16-byte word number
+                // Scratch  0x9_DE00    0x9DE0
+                // Ping     0x9_E000    0x9E00
+                // Pong     0x9_E200    0x9E20
+
+                let cfpa_word_number = {
+                    // Read the two versions. We do this with smaller buffers so
+                    // we don't need 2x 512B buffers to read the entire CFPAs.
+                    let mut ping_header = [0u32; 4];
+                    let mut pong_header = [0u32; 4];
+
+                    indirect_flash_read(
+                        &mut self.flash,
+                        0x9E00,
+                        core::slice::from_mut(&mut ping_header),
+                    )?;
+                    indirect_flash_read(
+                        &mut self.flash,
+                        0x9E20,
+                        core::slice::from_mut(&mut pong_header),
+                    )?;
+
+                    // Work out where to read the authoritative contents from.
+                    if ping_header[1] >= pong_header[1] {
+                        0x9E00
+                    } else {
+                        0x9E20
+                    }
+                };
+
+                // Read current CFPA contents.
+                let mut cfpa = [[0u32; 4]; 512 / 16];
+                indirect_flash_read(
+                    &mut self.flash,
+                    cfpa_word_number,
+                    &mut cfpa,
+                )?;
+
+                // Increment the monotonic version. The manual doesn't specify
+                // how the version numbers are compared or what happens if they
+                // wrap, so, we'll treat wrapping as an error and report it for
+                // now. (Note that getting this version to wrap _should_ require
+                // more write cycles than the flash can take.)
+                let new_version =
+                    cfpa[0][1].checked_add(1).ok_or(UpdateError::SecureErr)?;
+                cfpa[0][1] = new_version;
+                // Alter the boot setting. The boot setting (per RFD 374) is in
+                // the lowest bit of the 32-bit word starting at (byte) offset
+                // 0x100. This is flash word offset 0x10.
+                //
+                // Leave remaining bits undisturbed; they are currently
+                // reserved.
+                cfpa[0x10][0] &= !1;
+                cfpa[0x10][0] |= if slot == SlotId::A { 0 } else { 1 };
+                // The last two flash words are a SHA256 hash of the preceding
+                // data. This means we need to compute a SHA256 hash of the
+                // preceding data -- meaning flash words 0 thru 29 inclusive.
+                let cfpa_hash = {
+                    self.syscon
+                        .leave_reset(drv_lpc55_syscon_api::Peripheral::HashAes);
```

Comment:
Maybe add a comment here indicating that we call `enter_reset` during `lpc55-update-server`'s startup.

Reply:
Done!